phpunit/html/wp-html-tag-processor-test.php

```diff
@@ -151,6 +151,21 @@ public function test_attributes_parser_treats_slash_as_attribute_separator() {
 		$this->assertSame( 'test', $p->get_attribute( 'e' ), 'Accessing an existing e="test" did not return "test"' );
 	}
 
+	/**
+	 * @covers WP_HTML_Tag_Processor::get_attributes_by_prefix
+	 */
+	public function test_get_attributes_by_prefix_returns_set_of_prefixed_attributes() {
+		$p = new WP_HTML_Tag_Processor( '<div data-enabled class="test" data-test-id="14">Test</div>' );
```

Comment:
What should happen in these cases?

* `<div DATA-enabled class="test" data-test-id="14">Test</div>` and `$p->get_attributes_by_prefix( 'data-' );`
* `<div DATA-enabled class="test" data-test-id="14">Test</div>` and `$p->get_attributes_by_prefix( 'DATA-' );`

Reply:
Ah, nice one! 😄

Well, I'd say it should align with `get_attribute()`'s behavior. The "invariant" should be that if I do something like `$attrs = $p->get_attributes_by_prefix( $prefix );` and then call `$p->get_attribute()` for each of `$attrs`' keys, it shouldn't return `null` for any of them.

1. Let's assume `get_attribute()` is case-sensitive, i.e.

```php
$p = new WP_HTML_Tag_Processor( '<div DATA-enabled class="test" data-test-id="14">Test</div>' );
$p->get_attribute( 'DATA-enabled' ) === true;
$p->get_attribute( 'DATA-test-id' ) === null;

$p->get_attribute( 'data-enabled' ) === null;
$p->get_attribute( 'data-test-id' ) === '14';

$p->get_attribute( 'DATA-ENABLED' ) === null;
$p->get_attribute( 'DATA-TEST-ID' ) === null;
```

Then, IMO,
- `get_attributes_by_prefix( 'data-' )` should return `array( 'data-test-id' => '14')`
- `get_attributes_by_prefix( 'DATA-' )` should return `array( 'DATA-enabled' => true)`

2. Assuming `get_attribute()` is case-insensitive, i.e.

```php
$p = new WP_HTML_Tag_Processor( '<div DATA-enabled class="test" data-test-id="14">Test</div>' );
$p->get_attribute( 'DATA-enabled' ) === true;
$p->get_attribute( 'DATA-test-id' ) === '14';

$p->get_attribute( 'data-enabled' ) === true;
$p->get_attribute( 'data-test-id' ) === '14';

$p->get_attribute( 'DATA-ENABLED' ) === true;
$p->get_attribute( 'DATA-TEST-ID' ) === '14';
```

Then, IMO,
- `get_attributes_by_prefix( 'data-' )` should return `array( 'DATA-enabled' => true, 'data-test-id' => '14' )`
- `get_attributes_by_prefix( 'DATA-' )` should return `array( 'DATA-enabled' => true, 'data-test-id' => '14' )`

Does that make sense?

---

Reality-checking the behavior of `get_attribute()` is fun 😬 

```
$p = new WP_HTML_Tag_Processor( '<div DATA-enabled class="test" data-test-id="14">Test</div>' );
$p->next_tag();
$p->get_attribute( 'data-enabled' ) === null; // Okay, I guess we're case-sensitive.
$p->get_attribute( 'DATA-ENABLED' ) === null; // Yep, looks like.
$p->get_attribute( 'DATA-enabled' ) === null; // ???
```

Is this expected? 😬

Reply:
Oh golly, that's a bug. Here's a fix: https://github.com/WordPress/gutenberg/pull/46748

Reply:
Thank you very much for the fix! I just merged it 😄